pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,505 @@
+					proof_entry.current_count = 1;
+				} else {
+					proof_entry.current_count.ensure_add_assign(1)?;
+				}
+
+				Ok::<(), DispatchError>(())
+			}
+			InboundEntry::Message(_) => Err(Error::<T>::ExpectedMessageProofType.into()),
+		}
+	}
+
+	/// A pre-dispatch update involves increasing the `expected_proof_count` or
+	/// `current_count` of `self` with the one of `other`.
+	///
+	/// If a session ID change is detected, `self` is replaced completely by
+	/// `other`.
+	pub fn pre_dispatch_update(&mut self, other: Self) -> DispatchResult {
+		match (&mut *self, &other) {
+			// Message entries
+			(
+				InboundEntry::Message(self_message_entry),
+				InboundEntry::Message(other_message_entry),
+			) => {
+				if self_message_entry.session_id != other_message_entry.session_id {
+					*self = other;
+
+					return Ok(());
+				}
+
+				self_message_entry
+					.expected_proof_count
+					.ensure_add_assign(other_message_entry.expected_proof_count)?;
+
+				Ok(())
+			}
+			// Proof entries
+			(InboundEntry::Proof(self_proof_entry), InboundEntry::Proof(other_proof_entry)) => {
+				if self_proof_entry.session_id != other_proof_entry.session_id {
+					*self = other;
+
+					return Ok(());
+				}
+
+				self_proof_entry
+					.current_count
+					.ensure_add_assign(other_proof_entry.current_count)?;
+
+				Ok(())
+			}
+			// Mismatches
+			(InboundEntry::Message(_), InboundEntry::Proof(_)) => {
+				Err(Error::<T>::ExpectedMessageType.into())
+			}
+			(InboundEntry::Proof(_), InboundEntry::Message(_)) => {
+				Err(Error::<T>::ExpectedMessageProofType.into())
+			}
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Retrieves all stored routers and then filters them based
+	/// on the available routers for the provided domain.
+	pub(crate) fn get_router_ids_for_domain(
+		domain: Domain,
+	) -> Result<Vec<T::RouterId>, DispatchError> {
+		let stored_routers = Routers::<T>::get();
+
+		let all_routers_for_domain = T::RouterProvider::routers_for_domain(domain);
+
+		let res = stored_routers
+			.iter()
+			.filter(|stored_router| {
+				all_routers_for_domain
+					.iter()
+					.any(|available_router| *stored_router == available_router)
+			})
+			.cloned()
+			.collect::<Vec<_>>();
+
+		if res.is_empty() {
+			return Err(Error::<T>::NotEnoughRoutersForDomain.into());
+		}
+
+		Ok(res)
+	}
+
+	/// Calculates and returns the proof count required for processing one
+	/// inbound message.
+	pub(crate) fn get_expected_proof_count(
+		router_ids: &[T::RouterId],
+	) -> Result<u32, DispatchError> {
+		let expected_proof_count = router_ids
+			.len()
+			.ensure_sub(1)
+			.map_err(|_| Error::<T>::NotEnoughRoutersForDomain)?;
+
+		Ok(expected_proof_count.saturated_into())
+	}
+
+	/// Gets the message proof for a message.
+	pub(crate) fn get_message_proof(message: T::Message) -> Proof {
+		match message.get_proof() {
+			None => message
+				.to_proof_message()
+				.get_proof()
+				.expect("message proof ensured by 'to_message_proof'"),
```

Comment:
Because if not, I'm afraid here we can get a hash proof of a proof message instead of the message itself, which will be different: `hash(MessageProof(msg))= != hash(msg)`

Reply:
Removed from UTs.

Reply:
Only if we screw up the implementation in the LP message.

Reply:
Ah, okay. revisiting that code, the `to_proof_message().get_proof()` is only done by first checking that the message was not proof previously. So we're safe then 👍🏻 

Sorry for the noise here!

Reply:
No worries, we should look under every rock and try to make this logic as clear as possible!